File: proton-c/bindings/ruby/lib/core/disposition.rb
#--
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#++

module Qpid::Proton

  # Disposition records the current state and/or final outcome of a transfer.
  #
  # Every delivery contains both a local and a remote disposition. The local
  # disposition holds the local state of the delivery, and the remote
  # disposition holds the *last known* remote state of the delivery.
  #
  class Disposition

    include Util::Constants

    # Indicates the delivery was received.
    self.add_constant(:RECEIVED, Cproton::PN_RECEIVED)
    # Indicates the delivery was accepted.
    self.add_constant(:ACCEPTED, Cproton::PN_ACCEPTED)
    # Indicates the delivery was rejected.
    self.add_constant(:REJECTED, Cproton::PN_REJECTED)
    # Indicates the delivery was released.
    self.add_constant(:RELEASED, Cproton::PN_RELEASED)
    # Indicates the delivery was modified.
    self.add_constant(:MODIFIED, Cproton::PN_MODIFIED)

    # @private
    include Util::Engine

    attr_reader :impl

    # @private
    def initialize(impl, local)
      @impl = impl
      @local = local
      @data = nil
      @condition = nil
      @annotations = nil
    end

    # @private
    include Util::SwigHelper

    # @private
    PROTON_METHOD_PREFIX = "pn_disposition"

    # @!attribute section_number
    #
    # @return [Fixnum] The section number of the disposition.
    #
    proton_accessor :section_number

    # @!attribute section_offset
    #
    #  @return [Fixnum] The section offset of the disposition.
    #
    proton_accessor :section_offset

    # @!attribute failed?
    #
    # @return [Boolean] The failed flag.
    #
    proton_accessor :failed, :is_or_get => :is

    # @!attribute undeliverable?
    #
    # @return [Boolean] The undeliverable flag.
    #
    proton_accessor :undeliverable, :is_or_get => :is

    # Sets the data for the disposition.
    #
    # @param data [Codec::Data] The data.
    #
    # @raise [AttributeError] If the disposition is remote.
    #
    def data=(data)
      raise AttributeError.new("data attribute is read-only") unless @local
      @data = data
    end

    # Returns the data for the disposition.
    #
    # @return [Codec::Data] The data.
    #
    def data
      if @local
        @data
      else
        data_to_object(Cproton.pn_disposition_data(@impl))
      end
    end

    # Sets the annotations for the disposition.
    #
    # @param annotations [Codec::Data] The annotations.
    #
    # @raise [AttributeError] If the disposition is remote.
    #
    def annotations=(annotations)
      raise AttributeError.new("annotations attribute is read-only") unless @local
      @annotations = annotations
    end

    # Returns the annotations for the disposition.
    #
    # @return [Codec::Data] The annotations.
    #
    def annotations
      if @local
        @annotations
      else
        data_to_object(Cproton.pn_disposition_annotations(@impl))
      end
    end

    # Sets the condition for the disposition.
    #
    # @param condition [Codec::Data] The condition.
    #
    # @raise [AttributeError] If the disposition is remote.
    #
    def condition=(condition)
      raise AttributeError.new("condition attribute is read-only") unless @local
      @condition = condition
    end

    # Returns the condition of the disposition.
    #
    # @return [Codec::Data] The condition of the disposition.
    #
    def condition
      if @local
        @condition
      else
        condition_to_object(Cproton.pn_disposition_condition(@impl))
      end
    end

  end

end

File: proton-j/src/main/java/org/apache/qpid/proton/transport2/Disposition.java
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */

package org.apache.qpid.proton.transport2;

import java.util.List;

import org.apache.qpid.proton.codec2.DecodeException;
import org.apache.qpid.proton.codec2.DescribedTypeFactory;
import org.apache.qpid.proton.codec2.Encodable;
import org.apache.qpid.proton.codec2.Encoder;

public final class Disposition implements Encodable, Performative
{
    public final static long CODE = 0x0000000000000015L;

    public final static String DESCRIPTOR = "amqp:disposition:list";

    public final static Factory FACTORY = new Factory();

    private Role _role = Role.SENDER;

    private int _first;

    private int _last;

    private boolean _settled;

    private DeliveryState _state;

    private boolean _batchable;

    public Role getRole()
    {
        return _role;
    }

    public void setRole(Role role)
    {
        if (role == null)
        {
            throw new NullPointerException("Role cannot be null");
        }
        _role = role;
    }

    public int getFirst()
    {
        return _first;
    }

    public void setFirst(int first)
    {
        _first = first;
    }

    public int getLast()
    {
        return _last;
    }

    public void setLast(int last)
    {
        _last = last;
    }

    public boolean getSettled()
    {
        return _settled;
    }

    public void setSettled(boolean settled)
    {
        _settled = settled;
    }

    public DeliveryState getState()
    {
        return _state;
    }

    public void setState(DeliveryState state)
    {
        _state = state;
    }

    public boolean getBatchable()
    {
        return _batchable;
    }

    public void setBatchable(boolean batchable)
    {
        _batchable = batchable;
    }

    @Override
    public void encode(Encoder encoder)
    {
        encoder.putDescriptor();
        encoder.putUlong(CODE);
        encoder.putList();
        encoder.putBoolean(_role.getValue());
        encoder.putUint(_first);
        encoder.putUint(_last);
        encoder.putBoolean(_settled);
        if (_state == null)
        {
            encoder.putNull();
        }
        else
        {
            _state.encode(encoder);
        }
        encoder.end();
    }

    public static final class Factory implements DescribedTypeFactory
    {
        @SuppressWarnings("unchecked")
        public Object create(Object in) throws DecodeException
        {
            List<Object> l = (List<Object>) in;
            Disposition disposition = new Disposition();

            if (l.isEmpty())
            {
                throw new DecodeException("The first field cannot be omitted");
            }

            switch (6 - l.size())
            {

            case 0:
                disposition.setBatchable(l.get(5) == null ? false : (Boolean) l.get(5));
            case 1:
                disposition.setState((DeliveryState) l.get(4));
            case 2:
                disposition.setSettled(l.get(3) == null ? false : (Boolean) l.get(3));
            case 3:
                disposition.setLast((Integer) l.get(2));
            case 4:
                disposition.setFirst((Integer) l.get(1));
            case 5:
                disposition.setRole(Boolean.TRUE.equals(l.get(0)) ? Role.RECEIVER : Role.SENDER);
            }
            return disposition;
        }
    }

    @Override
    public String toString()
    {
        return "Disposition{" +
                "role=" + _role +
                ", first=" + _first +
                ", last=" + _last +
                ", settled=" + _settled +
                ", state=" + _state +
                ", batchable=" + _batchable +
                '}';
    }

    @Override
    public long getCode()
    {
        return CODE;
    }

    @Override
    public String getDescriptor()
    {
        return DESCRIPTOR;
    }
}